web_scraper/chicago_scraper.py

```diff
@@ -0,0 +1,155 @@
+elif current_os == "Darwin": 
+    service = Service(executable_path="drivers/chromedriver")  
+else:
+    raise Exception("Unsupported operating system")
+
+options = webdriver.ChromeOptions()
+options.add_argument("--headless=new")
+
+driver = webdriver.Chrome(options=options, service=service)
+url = "https://webapps1.chicago.gov/licensedcontractors/all"
+
+driver.get(url)
+table = driver.find_element(By.ID, 'paginatedTable')
+titles = table.find_elements(By.TAG_NAME, 'th')
+table_titles = [title.text for title in titles]
+
+table_content = []
+wait = WebDriverWait(driver, 5)
+current = 1
+max = 1
+
+table_rows = wait.until(expected_conditions.presence_of_all_elements_located((By.CSS_SELECTOR,'table#paginatedTable tbody tr')))
+
+for row in table_rows:
+    row_data = [item.text for item in row.find_elements(By.TAG_NAME, 'td')] 
+    table_content.append(row_data)
+
+while current < max: 
+    try :
+        next_bttn = driver.find_element(By.ID, 'paginatedTable_next')
+        next_bttn.find_element(By.TAG_NAME, 'a').click()
+        time.sleep(2)
+        current +=1
+        table_rows = wait.until(expected_conditions.presence_of_all_elements_located((By.CSS_SELECTOR,'table#paginatedTable tbody tr')))
+        for row in table_rows:
+            row_data = [item.text for item in row.find_elements(By.TAG_NAME, 'td')]             
+            table_content.append(row_data)
+    except Exception as e:
+        break
+
+data = pd.DataFrame(np.array(table_content), columns = table_titles)
+data['Tags'] = data['License Type'].apply(lambda x: "General" if "General Contractor" in x 
+                                        else "Electrical" if "Electrical Contractor" in x 
+                                        else "Elevator Mechanic" if "Elevator Mechanic Contractor" in x 
+                                        else "Mason" if "Mason Contractor" in x 
+                                        else "Plumbing" if "Plumbing Contractor" in x 
+                                        else "")
+
+for ind in data.index: #data.index is the label for each row, so for every one index in the dataframe
+    '''check if company_name exists in contractors table'''
+    addr = data['Address'][ind]
+    parts = addr.split(",")
+    if len(parts) == 3:
+        town = parts[1]
+    if len(parts) == 4:
+        town = parts[2]
+
+    license_number = data['License Number'][ind]
+    license_type = data['License Type'][ind]
+    license_expiration = data['License Expiration Date'][ind]
+    insurance_expiration = data['Insurance / Bond Expiration Date'][ind]
+    is_active = data['License Inactive?'][ind]
+
+    if license_expiration is None or license_expiration.strip() == "":
+        license_expiration = None
+    if insurance_expiration is None or insurance_expiration.strip() == "": 
+        insurance_expiration = None
+
+    list_company_names = supabase.table("Contractors").select("id").eq("company_name", data['Name'][ind]).execute()
```

Comment:
Some of these queries can be called once before the for loop so you don't have to call supabase so many times. Remember we want to limit the amount of times supabase gets called.

If you need to keep referencing this because you want to check if something needs to be created vs updated consider calling it once and then keeping track of the data as it gets added using something like a Set, Map or something else that you can easily append and check. 

Perhaps think about this after tackling the other comments.